dbms/src/DataStreams/NonJoinedBlockInputStream.cpp

```diff
@@ -243,33 +258,51 @@ size_t NonJoinedBlockInputStream::fillColumns(const Map & map,
     if (!position)
     {
         current_segment = index;
+
+        while (parent.partitions[current_segment]->spill)
+        {
+            current_segment += step;
+            if (current_segment >= map.getSegmentSize())
+            {
+                return rows_added;
+            }
+        }
+
         position = decltype(position)(
             static_cast<void *>(new typename Map::SegmentType::HashTable::const_iterator(map.getSegmentTable(current_segment).begin())),
             [](void * ptr) { delete reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(ptr); });
     }
 
+    if (current_segment >= map.getSegmentSize())
+    {
+        return rows_added;
+    }
     /// use pointer instead of reference because `it` need to be re-assigned latter
     auto it = reinterpret_cast<typename Map::SegmentType::HashTable::const_iterator *>(position.get());
     auto end = map.getSegmentTable(current_segment).end();
 
     for (; *it != end || current_segment + step < map.getSegmentSize();)
     {
-        if (*it == end)
+        if (*it == end || (parent.max_bytes_before_external_join && parent.partitions[current_segment]->spill))
```

Comment:
The spill check is different with L262, I think we should keep the same style, and since `parent.partitions[current_segment]->spill` is false by default even if spill is not enabled, I think just use the check in L262 is enough.

Reply:
> The spill check is different with L262, I think we should keep the same style, and since `parent.partitions[current_segment]->spill` is false by default even if spill is not enabled, I think just use the check in L262 is enough.

done.